hfile.c

```diff
@@ -573,6 +573,9 @@ static hFILE *hopen_fd(const char *filename, const char *mode)
 
 hFILE *hdopen(int fd, const char *mode)
 {
+#if defined HAVE_SETMODE && defined O_BINARY
+    if (setmode(fd, O_BINARY) < 0) return NULL;
+#endif
```

Comment:
@jmarshall, can you please clarify the intention here for binary files.  You state "The general policy in htslib is to open everything in binary mode", but that is precisely what this change does.  Do you mean that all opens should be "rb", "wb" or similar and specifying it at that point?  (Mostly we do that already of course.)  That only covers fopen style APIs though and not the Unix ```open``` function used in the test case.   The ```bgzf_dopen``` function should perhaps add "b" to the mode string before calling ```hdopen```, but this doesn't really seem to fit the "everything is binary" philosophy if we're only doing it for bgzf and not for SAM.

Stdin/stdout as you note are the exceptions which clearly need a ```setmode``` call to change, but again we're turning stdin/stdout to an hfile via the dopen call once more.